cherab/tools/raytransfer/roughconductor.pyx

```diff
@@ -0,0 +1,108 @@
+# European Commission - subsequent versions of the EUPL (the "Licence");
+# You may not use this work except in compliance with the Licence.
+# You may obtain a copy of the Licence at:
+#
+# https://joinup.ec.europa.eu/software/page/eupl5
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the Licence for the specific language governing permissions and limitations
+# under the Licence.
+#
+
+from numpy cimport ndarray
+from raysect.optical cimport Point3D, Vector3D, AffineMatrix3D, World, Ray, Spectrum, new_vector3d
+from libc.math cimport M_PI, sqrt
+from raysect.optical.material cimport RoughConductor
+cimport cython
+
+
+cdef class RToptimisedRoughConductor(RoughConductor):
+    """
+    A `RoughConductor` optimised for calculation of ray transfer matrix (geometry matrix).
+    The spectral array in this case contains ~ 10^5 - 10^6 spectral bins but the wavelengths for all of them are equal.
+    The Fresnel indeces are equal for all spectral bins, so the unnecessary calculations are avoided.
+    """
+
+    @cython.cdivision(True)
+    cpdef Spectrum evaluate_shading(self, World world, Ray ray, Vector3D s_incoming, Vector3D s_outgoing,
+                                    Point3D w_reflection_origin, Point3D w_transmission_origin, bint back_face,
+                                    AffineMatrix3D world_to_surface, AffineMatrix3D surface_to_world):
+
+        cdef:
+            double n, k
+            double ci
+            double wavelength
+            Vector3D s_half
+            Spectrum spectrum
+            Ray reflected
+
+        # outgoing ray is sampling incident light so s_outgoing = incident
+
+        # material does not transmit
+        if s_outgoing.z <= 0:
+            return ray.new_spectrum()
+
+        # ignore parallel rays which could cause a divide by zero later
+        if s_incoming.z == 0:
+            return ray.new_spectrum()
+
+        # calculate half vector
+        s_half = new_vector3d(
+            s_incoming.x + s_outgoing.x,
+            s_incoming.y + s_outgoing.y,
+            s_incoming.z + s_outgoing.z
+        ).normalise()
+
+        # generate and trace ray
+        reflected = ray.spawn_daughter(w_reflection_origin, s_outgoing.transform(surface_to_world))
+        spectrum = reflected.trace(world)
+
+        # evaluate lighting with Cook-Torrance bsdf (optimised)
+        wavelength = 0.5 * (spectrum.min_wavelength + spectrum.max_wavelength)
+        n = self.index.evaluate(wavelength)
+        k = self.extinction.evaluate(wavelength)
+        ci = s_half.dot(s_outgoing)
+        spectrum.mul_scalar(self._d(s_half) * self._g(s_incoming, s_outgoing) * self._fresnel_conductor(ci, n, k) / (4 * s_incoming.z))
+        return spectrum
+
+    @cython.cdivision(True)
+    cdef double _d(self, Vector3D s_half):
```

Comment:
I agree, however the names of these functions are exactly the same as in [RoughConductor](https://github.com/raysect/source/blob/master/raysect/optical/material/conductor.pyx) class from Raysect. It's better to keep the names identical in both classes.

Reply:
Yeah, I agree with @vsnever here. These are private methods so they are internal. They refer to the components of a RoughConductor equation. The terms in the equation are often denoted d, g, etc. So for anyone familiar with these equations it would be clear. This is a documentation problem and should be addressed elsewhere. The appropriate documentation in Raysect needs expanding.

Reply:
Ah, OK. I'd missed that they were already named this way by Raysect, so it makes sense to keep the same naming convention. 

However, it looks like they're actually exact copies of the `_d`, `_g` and `_g1` methods of `raysect.optical.material.conductor.RoughConductor`. Same for `_fresnel_conductor`. You're inheriting from this class, so these methods are already implemented in the base class. Unless I've missed something, there's no need to copy-paste these methods here. You're only overriding `evaluate_shading`, aren't you?